Buildings code rewrite

meased opened the pull request:
(Most likely) a prerequisite for #2532

This PR simplifies the buildings SQL query and moves all styling decisions out of SQL and into CSS (as the software gods intended :wink:).

No visual changes.

## SQL

Before, normal buildings and major building were queried separately with each query specifically omitting the results of the other. This made it difficult to redefine the categories (and very difficult to add a third category).

The SQL query now queries all buildings together, along with any fields needed to make styling decisions:

 ``` sql
      (SELECT
            way,
            building,
            amenity,
            aeroway,
            aerialway,
            tags->'public_transport' as public_transport
          FROM planet_osm_polygon
          WHERE building IS NOT NULL
            AND building != 'no'
            AND way_area > 1*!pixel_width!::real*!pixel_height!::real
          ORDER BY COALESCE(layer,0), way_area DESC
        ) AS buildings
```

## CSS

Now that the buildings class has access to all the fields needed, the CSS can be written in one block. As you can see, it is now very easy to reclassify buildings and add/remove the minor class:

``` css
  #buildings {
    [zoom >= 13] {
      polygon-fill: @building-low-zoom;
      polygon-clip: false;
      line-width: 0;
      [zoom >= 15] {
        polygon-fill: @building-fill;
        line-color: @building-line;
        line-width: .75;
        line-clip: false;
      }
      [amenity = 'place_of_worship'],
      [aeroway = 'terminal'],
      [aerialway = 'station'],
      [building = 'train_station'],
      [public_transport = 'station'] {
        polygon-fill: @building-major-fill;
        line-color: @building-major-line;
      }
    }
  }
```

jeisenbe commented:
Great work on the code!

I think it might be better with the major buildings a little more
prominent.

Could we try a color half-way between the old and new major building colors?

On Fri, Oct 5, 2018 at 11:34 AM meased <notifications@github.com> wrote:

> Fixes #2532
> <https://github.com/gravitystorm/openstreetmap-carto/issues/2532>
>
> This PR simplifies the buildings SQL query and moves all styling decisions
> out of SQL and into CSS (as the software gods intended 😉).
>
> Visual changes:
>
>    - Major buildings are not as dark.
>    - Adds a minor building category (lighter).
>
> SQL
>
> Before, normal buildings and major building were queried separately with
> each query specifically omitting the results of the other. This made it
> difficult to redefine the categories (and very difficult to add a third
> category).
>
> The SQL query now queries all buildings together, along with any fields
> needed to make styling decisions:
>
>      (SELECT
>            way,
>            building,
>            amenity,
>            aeroway,
>            aerialway,
>            tags->'public_transport' as public_transport
>          FROM planet_osm_polygon
>          WHERE building IS NOT NULL
>            AND building != 'no'
>            AND way_area > 1*!pixel_width!::real*!pixel_height!::real
>          ORDER BY COALESCE(layer,0), way_area DESC
>        ) AS buildings
>
> CSS
>
> Now that the buildings class has access to all the fields needed, the CSS
> can be written in one block. As you can see, it is now very easy to
> reclassify buildings and add/remove the minor class:
>
> #buildings {
>   [zoom >= 13] {
>     polygon-fill: @building-low-zoom;
>     polygon-clip: false;
>     line-width: 0;
>     [zoom >= 15] {
>       polygon-fill: @building-fill;
>       line-color: @building-line;
>       line-width: .75;
>       line-clip: false;
>     }
>     [amenity = 'place_of_worship'],
>     [aeroway = 'terminal'],
>     [aerialway = 'station'],
>     [building = 'train_station'],
>     [public_transport = 'station'] {
>       polygon-fill: @building-major-fill;
>       line-color: @building-major-line;
>     }
>     [building = 'garage'],
>     [building = 'garages'],
>     [building = 'carport'],
>     [building = 'shed'],
>     [building = 'greenhouse'],
>     [building = 'farm_auxiliary'],
>     [building = 'construction'],
>     [building = 'service'],
>     [building = 'ger'],
>     [building = 'ruins'] {
>       polygon-fill: @building-minor-fill;
>       line-color: @building-minor-line;
>     }
>   }
> }
>
> Test Renderings
>
> https://www.openstreetmap.org/#map=18/45.59116/-122.74872
>
> (Images line up so you can put them in browser tabs and switch tabs back
> and forth.)
>
> Current style on top.
>
> [image: sj_14b]
> <https://user-images.githubusercontent.com/14190496/46512926-21ad4880-c80b-11e8-9bd5-ce79a5a91950.png>
> [image: sj_14a]
> <https://user-images.githubusercontent.com/14190496/46512928-24a83900-c80b-11e8-9f8e-42587ae7b11d.png>
> [image: sj_15b]
> <https://user-images.githubusercontent.com/14190496/46512932-27a32980-c80b-11e8-8553-52a324fbff8a.png>
> [image: sj_15a]
> <https://user-images.githubusercontent.com/14190496/46512934-2a058380-c80b-11e8-8ed8-916e98be4b78.png>
> [image: sj_16b]
> <https://user-images.githubusercontent.com/14190496/46512936-2bcf4700-c80b-11e8-827a-5b0b0a33ada3.png>
> [image: sj_16a]
> <https://user-images.githubusercontent.com/14190496/46512937-2d990a80-c80b-11e8-9f73-165189c816d4.png>
> [image: sj_17b]
> <https://user-images.githubusercontent.com/14190496/46512939-2ffb6480-c80b-11e8-9501-9f829564db6a.png>
> [image: sj_17a]
> <https://user-images.githubusercontent.com/14190496/46512945-31c52800-c80b-11e8-81c8-7ef24abcc77d.png>
> [image: sj_18b]
> <https://user-images.githubusercontent.com/14190496/46512947-32f65500-c80b-11e8-927e-dd44db95486e.png>
> [image: sj_18a]
> <https://user-images.githubusercontent.com/14190496/46512950-34c01880-c80b-11e8-8054-73d2066c7941.png>
>
> https://www.openstreetmap.org/#map=16/45.5879/-122.5936
>
> [image: pa_16b]
> <https://user-images.githubusercontent.com/14190496/46513013-7ea8fe80-c80b-11e8-9c50-b90f7d78de4c.png>
> [image: pa_16a]
> <https://user-images.githubusercontent.com/14190496/46513014-7fda2b80-c80b-11e8-9e0e-80d8893805b4.png>
> [image: pa_17b]
> <https://user-images.githubusercontent.com/14190496/46513017-810b5880-c80b-11e8-8d43-5f445e98550f.png>
> [image: pa_17a]
> <https://user-images.githubusercontent.com/14190496/46513020-82d51c00-c80b-11e8-9e2c-d4ff7d16dffc.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/gravitystorm/openstreetmap-carto/pull/3426
> Commit Summary
>
>    - Standard text halo for fitness_center and fitness_station
>    - Merge branch 'master' of github.com:meased/openstreetmap-carto
>    - Merge https://github.com/gravitystorm/openstreetmap-carto
>    - xMerge https://github.com/gravitystorm/openstreetmap-carto
>    - Merge https://github.com/gravitystorm/openstreetmap-carto
>    - Rewrite buildings query, support for minor buildings
>    - Adjust colors
>    - Add comment
>    - Original building color
>
> File Changes
>
>    - *M* buildings.mss
>    <https://github.com/gravitystorm/openstreetmap-carto/pull/3426/files#diff-0>
>    (39)
>    - *M* project.mml
>    <https://github.com/gravitystorm/openstreetmap-carto/pull/3426/files#diff-1>
>    (31)
>
> Patch Links:
>
>    - https://github.com/gravitystorm/openstreetmap-carto/pull/3426.patch
>    - https://github.com/gravitystorm/openstreetmap-carto/pull/3426.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3426>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshM8MWXdk5DzEWYOHKTqYaib97KWCks5uhsVHgaJpZM4XJdL7>
> .
>

kocio-pl commented:
I like the idea of these complex changes. Did you however test the performance? We use SQL a lot to achieve reasonable speed.

kocio-pl commented:
Could you also test minor buildings darker, like half the way between the proposed color and standard buildings?

meased commented:
Minor buildings no change:
![0perc](https://user-images.githubusercontent.com/14190496/46559988-eeb89280-c8a6-11e8-86a7-3ab0be3b4243.png)

Minor buildings 4% lighter:
![4perc](https://user-images.githubusercontent.com/14190496/46559996-f841fa80-c8a6-11e8-8e13-ad3866b0837d.png)

Minor buildings 7% lighter (same as screen shots above):
![7perc](https://user-images.githubusercontent.com/14190496/46560015-08f27080-c8a7-11e8-9e03-024e7e8a7374.png)

meased commented:
Major buildings no change:
![maj2_20perc](https://user-images.githubusercontent.com/14190496/46560387-718e1d00-c8a8-11e8-8408-f063dde79103.png)
![maj_20perc](https://user-images.githubusercontent.com/14190496/46560395-79e65800-c8a8-11e8-9b99-139feea2edf1.png)

Major buildings 15% darker:
![maj2_15perc](https://user-images.githubusercontent.com/14190496/46560413-87034700-c8a8-11e8-9b0e-6b50c731fa4d.png)
![maj_15perc](https://user-images.githubusercontent.com/14190496/46560416-88cd0a80-c8a8-11e8-899f-e1756602c1bb.png)

Major buildings 10% darker (same as screenshots above):
![maj2_10perc](https://user-images.githubusercontent.com/14190496/46560432-95e9f980-c8a8-11e8-922a-c8745a483c91.png)
![maj_10perc](https://user-images.githubusercontent.com/14190496/46560434-971b2680-c8a8-11e8-8052-7fc896b6af4d.png)

As you can see, I was more concerned with fixing the dreadful "gastronomy in the airport" problem than I was with spotting churches in neighbourhoods.

meased commented:
> Did you however test the performance?

No.

Anyone know how to run benchmarks in docker?

jeisenbe commented:
Would it make sense to lighten major buildings at higher zoom levels? At
mid-zoom it is very helpful that the airport terminal building is obvious,
and that’s when it is useful for places of worship etc.

But at zoom 18 there is no reason for the major building to stay so dark.

Would that be too complicated ?
On Sat, Oct 6, 2018 at 6:45 AM meased <notifications@github.com> wrote:

> Did you however test the performance?
>
> No.
>
> Anyone know how to run benchmarks in docker?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3426#issuecomment-427507419>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshCBb_Z9SgjZ3WC1j5qpWh9E_4cxRks5uh9LngaJpZM4XJdL7>
> .
>

meased commented:
That's not that complicated. Worth a test rendering at least.

kocio-pl commented:
I think both darker versions (4% for minor and 15% for major) work better. Since buildings are very popular map feature, we need much more testing in different areas (backgrounds). 

I like the idea of minor buildings instead of problems with ruins icon, especially for castles (see #331).

Unfortunately we have no performance framework (see #1287). However we can at least try to deploy it on some real server - @rrzefox, could you measure the effect of replacing some SQL code with Carto CSS here?

matthijsmelissen commented:
See also #68 and #490.

I'm not sure if I really see the advantage of rendering minor buildings lighter.

mboeringa commented:
> I'm not sure if I really see the advantage of rendering minor buildings lighter.

That surprises me. To me it makes utter sense to render relatively insignificant non-living quarters like small garden sheds and garages lighter.

It might actually stimulate good tagging practices in that people actually start to distinguish them in tagging, and add **building=shed** instead of just generic **building=yes**.

Tomasz-W commented:
@meased Can you test some lighter outline for minor buildings? It looks a little bit like some holes in ground currently.

mboeringa commented:
> @meased Can you test lighter outline for minor buildings? It looks a little bit like some holes in ground currently.

I think that is mainly a consequence of the relatively dark grey of the **landuse=residential**, which appears just a bit darker than the minor buildings. So either slightly lightening up of landuse=residential, or slightly darkening the minor buildings fill, might help.

meased commented:
> I think that is mainly a consequence of the relatively dark grey of the landuse=residential

Exactly.

I choose landuse=residential and airports for examples because they represent the worst cases for how light you can go on one end and how dark you can go on the other. There is not much space to play with in the middle. The minor buildings above would have looked better on bare background:
![base_minor](https://user-images.githubusercontent.com/14190496/46574394-47476880-c957-11e8-9ddd-84e39330754e.png)

kocio-pl commented:
> It might actually stimulate good tagging practices in that people actually start to distinguish them in tagging, and add building=shed instead of just generic building=yes.

I also believe it's more pleasant, less cluttering and more useful at the same time, so it's a triple win for me.

There are a lot of auxiliary buildings, but still most of them is tagged with `building=yes` (and only about 50 mln out of 300 mln of all the buildings has any other value!). I don't see how could we attract people to more precise tagging other than showing them a bit different as a "minor" super-class.

![taghistory 55](https://user-images.githubusercontent.com/5439713/46576631-e78ca400-c9ce-11e8-9ba4-3471980b8064.png)
![taghistory 56](https://user-images.githubusercontent.com/5439713/46576633-e9566780-c9ce-11e8-86b4-2d99ef8ad3fe.png)

@meased I see that you have added carpark type, but omitted collapsed - was there any reason for this? I have tried to stay above 10k limit, but we don't have to stick to it, there are also some other types that make sense for me, what do you think about it?

![taghistory 57](https://user-images.githubusercontent.com/5439713/46576662-cbd5cd80-c9cf-11e8-9718-8f623e40b803.png)

jragusa commented:
Please also consider underground #552 and roof #2475 buildings to get a consistent rendering.

meased commented:
I got the original minor buildings list from #2532. I was avoiding underground (because some people don't want it rendered at all) and roof (because #2475 is different than just rendering them lighter). But until those other issues are sorted out I don't see why they wouldn't qualify as minor buildings in the meantime.

@kocio-pl: Strangly, collapsed isn't on the [wiki page](https://wiki.openstreetmap.org/wiki/Key:building). I included it anyway because the usage is very high.

The list of minor buildings is now: 

``` css
    [building = 'roof'],
    [building = 'construction'],
    [building = 'damaged'],
    [building = 'collapsed'],
    [building = 'underground'],
    [building = 'service'],
    [building = 'slurry_tank'],
    [building = 'shed'],
    [building = 'garage'],
    [building = 'garages'],
    [building = 'farm_auxiliary'],
    [building = 'carport'],
    [building = 'barn'],
    [building = 'stable'],
    [building = 'cowshed'],
    [building = 'greenhouse'],
    [building = 'ger'],
    [building = 'ruins'] {
```

meased commented:
Playing with @jeisenbe's idea of variable major building darkness:

![a14](https://user-images.githubusercontent.com/14190496/46584852-614b7e80-ca1d-11e8-9141-24941f38037a.png)
![a15](https://user-images.githubusercontent.com/14190496/46584853-63154200-ca1d-11e8-946a-08df774bce69.png)
![a16](https://user-images.githubusercontent.com/14190496/46584854-65779c00-ca1d-11e8-8e6b-312ceed4ffbe.png)
![a17](https://user-images.githubusercontent.com/14190496/46584855-67d9f600-ca1d-11e8-8ecc-02ee6d149b03.png)


![c14](https://user-images.githubusercontent.com/14190496/46584857-6b6d7d00-ca1d-11e8-8a5b-39e612d66464.png)
![c15](https://user-images.githubusercontent.com/14190496/46584859-6e686d80-ca1d-11e8-9af8-bfe1d1454b83.png)
![c16](https://user-images.githubusercontent.com/14190496/46584860-70cac780-ca1d-11e8-897d-716ea34ea9ae.png)
![c17](https://user-images.githubusercontent.com/14190496/46584861-71fbf480-ca1d-11e8-8546-977b22328ce7.png)

Tomasz-W commented:
About test renderings above: I find darkness of major buildings on high zoom levels definetely too low. When I looked at them with "fresh eye" I had to search for major buildings, but they should be distinguishable intuitively. Just to be clear: I very like the idea proposed in this PR, it just need some tuning IMO.

HolgerJeromin commented on buildings.mss:
```diff
@@ -12,29 +15,41 @@
+    [building = 'stable'],
+    [building = 'cowshed'],
+    [building = 'greenhouse'],
+    [building = 'ger'],
```
What is that?

I miss =hut and perhaps =cabin in the list.

kocio-pl replied:
You mean ger? See https://en.wikipedia.org/wiki/Yurt

kocio-pl commented:
What about construction or ruins?

dieterdreist commented:
sent from a phone

> On 9. Oct 2018, at 03:58, kocio-pl <notifications@github.com> wrote:
> 
> What about construction or ruins?


I’m undecided, they are not building types but they work like elsewhere, e.g. highway=construction, construction=primary

And they are at least buildings

kocio-pl commented:
@meased Variable darkness is smart trick and it can work for me. More tests would be useful to make sure, because buildings are very popular. 

The most important think for now is deciding what the list should contain. I like the idea of tagging `building=construction/ruins/collapsed/damaged` + `construction/ruins/collapsed/damaged=*`, but I'm not sure what should we render then - their uncomplete state as lighter or just their type.

***

Example of gers (yurts/tents) being less visible than regular buildings (in this area some of them are not tagged properly, so the difference in rendering is evident), using current code:

https://www.openstreetmap.org/#map=18/47.93434/106.91192

![ft9dvqo1](https://user-images.githubusercontent.com/5439713/46779392-dcf84480-cd17-11e8-8909-b0419689c51d.png)

rrzefox commented:
> However we can at least try to deploy it on some real server - @rrzefox, could you measure the effect of replacing some SQL code with Carto CSS here?

I will gladly deploy this once it has reached "beta" state, i.e. coloring has been decided.
I seriously doubt this has any measureable performance impact.

Adamant36 commented:
@rrzefox, since we are on the subject of performance, do you know what impact there would be from creating more stylesheets with what is currently in the amenity file?

Tomasz-W commented:
@meased Can you try with 17% for major and 5% for minor due to some users questions about further tests? It would be good to see some new locations to have a 'fresh view' on it.

meased commented:
No problem, I'll do more testing this weekend.

kocio-pl commented:
I think deploying this idea will help people realize that there are many different building types and start adding them more often. In the looooong term this might lead to render `building=yes` as kind of minor/incomplete. So this is much more to this change than meets the eye!

I believe in such incremental, but more systematic changes. For example making more categories for POIs, as I did purely for making this style easier to customize, has led to making gastronomy icons orange, which may help to avoid clutter by starting rendering them as dots.

dieterdreist commented:
sent from a phone

> On 12. Oct 2018, at 19:29, kocio-pl <notifications@github.com> wrote:
> 
> I think deploying this idea will help people realize that there are many different building types and start adding them more often. In the looooong term this might lead to render building=yes as kind of minor/incomplete. So this is much more to this change than meets the eye!


+1, well said, the times that mapping projects called themselves proudly building=yes should be gone ;-)

meased commented:
I have removed `collapsed` and added `hut` and `cabin`.

I have left `ruins` off due to similarity to `collapsed`.

Still undecided on `construction`. I feel it almost warrants a new style altogether, which would be better implemented in a separate PR, so I've left it off for now.

There is a lot going on in this PR so I am thinking about making this PR only about adding minor buildings (and the query rewrite) and making a separate PR later for lightening major buildings. I think this would help get things done faster as we will be able to focus on one thing at a time. Unless anyone objects, I'll update the repository and change the title of this PR.

meased commented:
More testing (at 7% lighter):

https://www.openstreetmap.org/#map=17/45.58231/-122.73927
![sj_b](https://user-images.githubusercontent.com/14190496/46919347-138dc380-cf93-11e8-9d5b-24cddb32e8d9.png)
![sj_a](https://user-images.githubusercontent.com/14190496/46919352-1b4d6800-cf93-11e8-9fcc-10983fdd8cb7.png)

https://www.openstreetmap.org/#map=17/45.15308/-122.84857
![wb_b](https://user-images.githubusercontent.com/14190496/46919354-256f6680-cf93-11e8-9055-08c32454ae5c.png)
![wb_a](https://user-images.githubusercontent.com/14190496/46919355-27d1c080-cf93-11e8-8295-8095f8944cce.png)

https://www.openstreetmap.org/#map=18/45.46846/-122.73448
![hv_b](https://user-images.githubusercontent.com/14190496/46919359-315b2880-cf93-11e8-93fa-32d3c5d61aa5.png)
![hv_a](https://user-images.githubusercontent.com/14190496/46919360-33bd8280-cf93-11e8-97bc-69833e73af7e.png)

https://www.openstreetmap.org/#map=17/45.50819/-122.64908
![la_b](https://user-images.githubusercontent.com/14190496/46919361-3ae49080-cf93-11e8-8807-a124d59a3391.png)
![la_a](https://user-images.githubusercontent.com/14190496/46919366-446df880-cf93-11e8-889d-16294ebbc271.png)

![aly_b](https://user-images.githubusercontent.com/14190496/46919369-4f288d80-cf93-11e8-9599-150f017bc438.png)
![aly_a](https://user-images.githubusercontent.com/14190496/46919370-50f25100-cf93-11e8-9fa0-bfe522e3a3c2.png)

meased commented:
@Tomasz-W wanted to see 5% lighter:

0%
![la1_0p](https://user-images.githubusercontent.com/14190496/46919454-9400f400-cf94-11e8-898d-e7791f2809e6.png)

5%
![la1_5p](https://user-images.githubusercontent.com/14190496/46919455-96fbe480-cf94-11e8-9b85-ffc123872008.png)

7%
![la1_7p](https://user-images.githubusercontent.com/14190496/46919457-98c5a800-cf94-11e8-8662-2e291b92f045.png)

0%
![la2_0p](https://user-images.githubusercontent.com/14190496/46919467-ad09a500-cf94-11e8-80ae-1b54e5a5510d.png)

5%
![la2_5p](https://user-images.githubusercontent.com/14190496/46919469-b09d2c00-cf94-11e8-883f-47b47f1a3942.png)

7%
![la2_7p](https://user-images.githubusercontent.com/14190496/46919471-b3981c80-cf94-11e8-9177-fbf1123139ee.png)

Tomasz-W commented:
Thanks! As 5% is too dark to distinguish them fastly, but 7% is minimally too light (too prominent) on residential areas, I would go with middle one of 6%.

meased commented:
6% version of the last image:
![la2_6p](https://user-images.githubusercontent.com/14190496/46919842-3ff90e00-cf9a-11e8-9d71-87f017a21491.png)

jeisenbe commented:
Looks good on residential landuse.

Can we try an example with minor buildings on Religious landuse?
On Mon, Oct 15, 2018 at 2:19 AM meased <notifications@github.com> wrote:

> 6% version of the last image:
> [image: la2_6p]
> <https://user-images.githubusercontent.com/14190496/46919842-3ff90e00-cf9a-11e8-9d71-87f017a21491.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3426#issuecomment-429644375>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshNMFTgzx5RQ88T8qrGWGs94PD4yuks5uk3IjgaJpZM4XJdL7>
> .
>

Adamant36 commented:
@meased, can you try a couple of tests with 6% and the new yellowish school grounds color?

kocio-pl commented:
I think it's important to test it on different backgrounds where one can find buildings (for example `landuse=garages` or `landuse=commercial`), not just the bare ground. They will look different there, so it's too early for precise shade tuning.

kocio-pl commented:
@meased My proposition is to start with just the code restructuring as a separate PR, so testing would be easier and less error prone, and next making another PR based on this which will be focused only on making minor category and visual testing the colors.

meased commented:
> My proposition is to start with just the code restructuring as a separate PR

I think I'll do this.

meased commented:
Done.

This PR is now just a rewrite of the buildings SQL and style sheet. There should be no visual differences. I have tested this in Kosmtik by flipping the overlay of the standard map on and off and I can't see any difference on any zoom level.

kocio-pl commented:
Great! Since there is a change from SQL code to CSS code, there might be some performance difference. @rrzefox Could you test this difference?

kocio-pl commented:
Well, this statement was not accurate - it's rather simpler both SQL and CSS code, so there should be no problems with performance. I will just wait with this until v4.16.0 is out (I plan to do this 19.10) and probably merge this PR soon after that. Then we can start testing minor buildings PR.

kocio-pl commented:
BTW: making major buildings lighter would also fix #3071.

kocio-pl commented:
@meased What about further tuning? Do you plan to get back to the proposed minor and major buildings changes?

meased commented:
The problem here is that buildings need to simultaneously serve as a background for our lightest foreground (I believe that is gastromony `#C77400`) and as a foreground to our darkest landusage (which I believe is religous `#d0d0d0`, but residential is also a problem), all while having three brightnesses of minor, normal, and major which are all distinguishable from each other.

I'm not convinced that all of these criteria can be met, hence the inactivity.

Regardless, I could start a PR so that these problems could be investigated.

matthijsmelissen commented:
That's the kind of thinking we need! Feel free to propose dropping/altering any of these criteria if you think it would make sense.

By the way, just curious, what's your username on openstreetmap.org?

meased commented:
> what's your username on openstreetmap.org?

[cowdog](https://www.openstreetmap.org/user/cowdog)

Tomasz-W commented:
Please remember about discussed half-transparency for some buildings (https://github.com/gravitystorm/openstreetmap-carto/issues/552#issuecomment-446540473).
